docs/SharedBrokerPool.md

```diff
@@ -0,0 +1,43 @@
+---
+id: shared-broker-pool
```

Comment:
What do you think about re-using the term "connection pool" for this, rather than the term "broker pool"?

It's not exactly the same thing, but it's very close, and it might make it easier for people to understand what this does and why they might want to do it. I only want to invent new terms if using existing ones misleads and causes more problems than they solve.

Reply:
Fair point. 

But I suspect then that it would make sense to completely rename things also in the code to stay consistent?

Reply:
> But I suspect then that it would make sense to completely rename things also in the code to stay consistent?

I instinctually think you are correct, as it rarely makes sense to call things one thing externally and another thing internally, but in this case I'm not sure. The reason I think BrokerPool makes sense internally is because it is not just holding connections, it is also responsible for holding metadata and mapping between node ids and connections (through the broker abstraction). So internally, it makes sense to call it a BrokerPool.

We initially called it that because we wanted to have a method that allowed us to execute something on any connected broker (`.withBroker((fn: { nodeId: string, broker: Broker }) => Promise<void>)`), which made it sound like it was a pool of brokers that we could borrow from.

Externally, however, the point is to communicate that this is a way to share connections between different instances, which is why Connection Pool is a good name. No one that hasn't dug into the innards of KafkaJS has ever heard of a BrokerPool, and I don't think you intuitively understand that it's a way to share connections unless you have experience developing KafkaJS itself.

I'm really not 100% sure. I am quite sure that I like the name "Connection Pool" externally - certainly more than I like exposing the name "Broker Pool". But I'm not sure whether or not I want to rename everything internally.

Reply:
I tried just touching the external API to use "connection pool", I think I agree: This reads easier and makes intuitively sense. Keeping the existing internal BrokerPool is fine, and in fact makes it more obvious that one cannot just create one of these things manually -- one must go through the exposed API on the client.

See https://github.com/tulios/kafkajs/pull/977/commits/a02f4557e8b12b896de28a9e9a79ff9422f17838